accepted/future-releases/enhanced-enums/feature-specification.md

```diff
@@ -152,20 +155,18 @@ It’s currently a compile-time error for a class to implement, extend or mix-in
 
 Because we want to allow interfaces and mixins that are intended to be applied to `enum` declarations, and therefore to assume `Enum` to be a superclass, we loosen that restriction to:
 
-- It's a compile-time error if a *non-abstract* class has `Enum` as a superinterface (directly or transitively) unless it is the corresponding class of an `enum` declaration.
+- It's a compile-time error if a *non-abstract* class has `Enum` as a superinterface (directly or transitively) unless it is the corresponding class of an `enum` declaration. _(Abstract interfaces and mixins implementing `Enum` are allowed, but only so that they can be used by `enum` declarations, they can never be used to create an instance which implements `Enum`, but which is not an enum value.)_
 - It is a compile-time error if a class implements, extends or mixes-in the class or interface introduced by an `enum` declaration. _(An enum class can’t be used as a mixin since it is not a `mixin` declaration and the class has a superclass other than `Object`, but we include “mixes-in” for completeness.)_
 - It's a **compile-time error** if a `class` or `mixin` declaration has `Enum` as a superinterface and the interface of the declarations contains an instance member with the name `values`, whether declared or inherited. _If any concrete class implements this interface, it will be an `enum` declaration class, and then the `values` member would conflict with the static `values` constant getter that is automatically added to `enum` declaration classes. Such an instance `values` declaration is either useless or wrong, so we disallow it entirely._
-- It's a compile-time error if a `class`, `mixin` or `enum` declaration has `Enum` as a superinterface, and it declares a non-abstract instance member named `index`. _That member would override the `index` getter inherited from `Enum`, and we currently do not allow that._
+- It's a compile-time error if a `class`, `mixin` or `enum` declaration has `Enum` as a superinterface (trivially true for `enum` declarations), and that declaration contains non-abstract instance member declaration named `index`, `hashCode` or `==` (an `operator ==` declaration). _That `index` member would override the `index` getter inherited from `Enum`, and we currently do not allow that. The `hashCode` and `operator==` declarations would prevent the enum class from having "primitive equality", and we want to ensure that enums can be used in switches._
```

Comment:
A couple of changes would fit here, for consistency:
```suggestion
- It's a compile-time error if a `class`, `mixin` or `enum` declaration has `Enum` as a superinterface (*trivially true for `enum` declarations*), and that declaration contains a concrete instance member declaration named `index`, `hashCode` or operator `==`. _That `index` member would override the `index` getter inherited from `Enum`, and we currently do not allow that. The `hashCode` and operator `==` declarations would prevent the enum class from having "primitive equality", and we want to ensure that enums can be used in switches._
```

However, there is another issue: This item repeats rules that have already been stated (in line 134 and 136), and this normally implies that the duplicate should be commentary. A similar issue arises for `values` in line 160.

Reply:
This actually *doesn't* duplicate the earlier errors.

This error only apply to *non-abstract declarations* of `index`/`hashCode`/`==` in a `class`/`mixin`/`enum` which implements `Enum`.

The errors in lines 134/136 apply to an `enum` declaration which declares or inherits `index`/`hashCode`/`==`.

So, the case of an `enum` with a mixin application of a mixin *not* implementing `Enum`, but declaring a non-abstract `index`/`hashCode`/`==`, is only covered by the earlier errors, not this one.

And that's probably an oversight, because we don't cover the case of:
```dart
class C {
  int get index => 0;
}
abstract class D extends C implements Enum {}
```
either.

It's actually not particularly *important* that a `class` implementing `Enum` has a concrete implementation of anything, because there is currently no way to ever get that implementation into an `enum` declaration.
(This error catches the class itself declaring `index`, etc., and if it's declared in a superclass, then the class has a superclass other than `Objecf` and cannot be used as a mixin.)

So, how strict should we be about classes implementing `Enum` and having implementations. Those implementations are guaranteed dead code, so we could choose to refuse it entirely.
Or we could just not care (status quo), since you'll probably figure it out if you ever try to use the class for anything except `implements` its interface.

Reply:
I actually prefer "non-abstract" to "concrete" because being `abstract` is the actual language term.

Reply:
The language specification has 58 occurrences of `concrete` and zero occurrences of 'non-abstract', so we do have prior art to support the notion that 'concrete' is well-defined. We could switch over, of course, but 'non-abstract' does seem somewhat bulky and hard to pronounce.

Reply:
The
> instance member declaration named `index`, `hashCode` or operator `==`

phrase doesn't work for me. There is no "instance member declaration named ... operator `==`".
It's either *named* `==` or it's an `operator ==` declaration.

Reply:
NB: I edited several 'suggestions' above to fit the results of the discussion at this point. Added `[Edit: ...]` there.

> The errors in lines 134/136 apply to an `enum` declaration

Indeed, I forgot about that distinction even though it came up earlier in discussions with the UniPro folks. ;-)

I changed the edit suggestion such that the overlap is reduced: The rule here will then apply to classes and mixins, not to enum declarations.

However, these error specifications should occur right next to each other, it is too hard to read the document when they are so far apart from each other.

> So, the case of an `enum` with a mixin application of a mixin not implementing `Enum`, but declaring a non-abstract `index`/`hashCode`/`==`, is only covered by the earlier errors, not this one.

Right, and we do have cases where line 161 applies and line 134/136 do not:

```dart
mixin M on Enum {
  int get index => 0; // Error.
}

abstract class B implements Enum {
  int get index => 0; // Error.
}

class C {
  int get index => 0;
}
abstract class D extends C implements Enum {} // Error.
```

I suggested that we should use the phrase '.. C has a concrete instance member `m`..', meaning that the superclass chain of `C` contains an element `B` (which could be `C` or some other class, possibly a mixin application) such that `B` declares a concrete instance member named `m`.

There is a bit of a conceptual clash here, because 'has'/'declares' is usually used with interfaces, that is, with member signatures and not actual declarations, but I think it is quite readable and not ambiguous in practice.